Add support for folders in flows, readModel, writeModel. thenativeweb/wolkenkit#150

jbeaudoin11 opened the pull request:
Updated tests
Cleaned commits

I added .editorconfig which is really helpful since i use a different config in my vs code settings. I can remove it if you want me to, but you guys should start using it.

All test pass, but i haven't tested it with a real project since i'm not sure how to include my fork.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thenativeweb/wolkenkit-application?pullRequest=2) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thenativeweb/wolkenkit-application?pullRequest=2) before we can accept your contribution.<br/><hr/>**Jimmy Beaudoin** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thenativeweb/wolkenkit-application?pullRequest=2) it.</sub>

goloroden commented on .editorconfig:
```diff
@@ -0,0 +1,9 @@
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = false
```
This value should be `true`, to be consistent with our usual editor settings. Could you please change that?

(Oh, and by the way, this line needs a final newline 😂)

goloroden commented on src/getDirectoryTree.js:
```diff
@@ -0,0 +1,69 @@
+
+const access = promisify(fs.access);
+
+const directoryTree = require('directory-tree');
```
We typically sort our require calls in a slightly different way: Three blocks with a single `const`, for built-in modules, for npm modules, and for local files. Hence, you'd basically did everything correctly, I just wanted to ask whether you could swap the linel that loads the `directory-tree` module and the one that promisifies the `fs.access` function?

goloroden commented on src/getDirectoryTree.js:
```diff
@@ -0,0 +1,69 @@
+
+const directoryTree = require('directory-tree');
+
+const MAX_TREE_TRANSFORM_DEPTH = {
```
We don't use uppercasing or underscores for constants. Could you please turn this into camel case as well?

goloroden commented on src/getDirectoryTree.js:
```diff
@@ -0,0 +1,69 @@
+  // flows/flow[.js, /index.js]
+  flows: 1,
+
+  // readModel/api/projection[.js, /index.js]
```
`api` should be `lists`, and `projection` should be `model`, to be consistent with the naming.

goloroden commented on src/getDirectoryTree.js:
```diff
@@ -0,0 +1,69 @@
+  // writeModel/context/aggregate[.js, /index.js]
+  writeModel: 2
+};
+const transformTree = function (nodes, maxDepth = 0, currentDepth = 0) {
```
Here, a newline would be nice.

goloroden commented on src/getDirectoryTree.js:
```diff
@@ -0,0 +1,69 @@
+
+    const maxTreeTransformDepth = MAX_TREE_TRANSFORM_DEPTH[name];
+
+    if (maxTreeTransformDepth) {
```
I suppose that this code works (as all the things you did are very well done 😊), but maybe there is a way to make it a little bit more readable, i.e. remove some of the nesting, and the `else` blocks? I don't have a specific suggestion here, I'm just thinking out loud…

jbeaudoin11 replied:
Would you like me to add comments ? I know it's not that trivial so it could help

goloroden replied:
I think it's fine as it is 👍

goloroden commented on src/validateEntries.js:
```diff
@@ -7,6 +7,8 @@ const validateStructure = async function ({ entries }) {
     throw new Error('Entries are missing.');
   }
 
+  console.log(entries)
```
This is probably a leftover of some debugging, isn't it? I suppose this can be removed…

jbeaudoin11 commented:
Alright, I made all changes you asked.
I added comments and changed some names to the `getDirectoryTree` algo, let me know if it's better.

>PPS: Please note that you also need to sign the CLA before we are able to merge this.

I had problem with the CLA, my git wasn't using the proper email adresse. Had to rebase and edit my commit email, that's why there is a force push.

Actually would you like me to squash my commits together ? It's easier to read the changes you asked in a separate commit that's why i didn't.

>PS: Have you tried whether it also works correctly when there is more than just an index.js file in a folder? Maybe it would make sense to add an aggregate, a list and a flow with multiple files as well?

I made sure all test pass with those modifications. There is already other files in the `folderAggregate`.

goloroden commented:
From my side, everything is now ready to go 😊 

You don't need to squash, I will do this when merging things. But thanks for asking!

One last thing: You didn't add yourself as a contributor to the `package.json` file – was this by intention or by accident?

As soon as I hear from you, I will merge this PR 😊

jbeaudoin11 commented:
> One last thing: You didn't add yourself as a contributor to the package.json file – was this by intention or by accident?

Oops, forgot about that ! Thanks for the reminder, i just added my self.

You have my go for the merge.

jbeaudoin11 commented:
>Could you please remove the dist folder from your PR
Also, forgot about that one
